backends/ebpf/psa/externs/ebpfPsaTableImplementation.cpp

```diff
@@ -206,4 +207,444 @@ void EBPFActionProfilePSA::applyImplementation(CodeBuilder* builder, cstring tab
+    builder->endOfStatement(true);
+    builder->blockEnd(true);
+
+    // 5. Execute action.
+
+    builder->emitIndent();
+    builder->appendFormat("if (%s != NULL) ", asValueName.c_str());
+    builder->blockStart();
+
+    emitAction(builder, asValueName, actionRunVariable);
+
+    builder->blockEnd(false);
+    builder->append(" else ");
+
+    builder->blockStart();
+    builder->target->emitTraceMessage(builder,
+        "ActionSelector: member not found, executing implicit NoAction");
+    builder->emitIndent();
+    builder->appendFormat("%s = 0", program->control->hitVariable.c_str());
+    builder->endOfStatement(true);
+    if (!actionRunVariable.isNullOrEmpty()) {
+        builder->emitIndent();
+        builder->appendFormat("%s = 0", actionRunVariable.c_str());  // set to NoAction
+        builder->endOfStatement(true);
+    }
+    builder->blockEnd(true);
+
+    msg = Util::printf_format("ActionSelector: %s applied", instanceName.c_str());
+    builder->target->emitTraceMessage(builder, msg.c_str());
+}
+
+EBPFHashAlgorithmPSA::ArgumentsList EBPFActionSelectorPSA::unpackSelectors() {
+    EBPFHashAlgorithmPSA::ArgumentsList result;
+    for (auto s : selectors) {
+        result.emplace_back(s->expression);
+    }
+    return result;
+}
+
+EBPFActionSelectorPSA::SelectorsListType
+EBPFActionSelectorPSA::getSelectorsFromTable(const EBPFTablePSA * instance) {
+    SelectorsListType ret;
+
+    for (auto k : instance->keyGenerator->keyElements) {
+        auto mkdecl = program->refMap->getDeclaration(k->matchType->path, true);
+        auto matchType = mkdecl->getNode()->to<IR::Declaration_ID>();
+
+        if (matchType->name.name == "selector")
+            ret.emplace_back(k);
+    }
+
+    return ret;
+}
+
+void EBPFActionSelectorPSA::registerTable(const EBPFTablePSA * instance) {
+    if (table == nullptr) {
+        selectors = getSelectorsFromTable(instance);
+        emptyGroupAction =
+                instance->table->container->properties->getProperty("psa_empty_group_action");
+        groupsMapSize = instance->size;
+    } else {
+        verifyTableSelectorKeySet(instance);
+        verifyTableEmptyGroupAction(instance);
+
+        // Documentation says: The number of groups may be at most the size of the table
+        //     that is implemented by the selector.
+        // So, when we take the max from both tables, this value will not be conformant
+        // with specification because for one table there will be available more groups
+        // than possible entries. We have to use min.
+        groupsMapSize = std::min(groupsMapSize, instance->size);
+    }
+
+    EBPFTableImplementationPSA::registerTable(instance);
+}
+
+void EBPFActionSelectorPSA::verifyTableSelectorKeySet(const EBPFTablePSA * instance) {
+    bool printError = false;
+    auto foreignSelectors = getSelectorsFromTable(instance);
+
+    if (selectors.size() == foreignSelectors.size()) {
```

Comment:
this kind of checks should probably be done in the PSA architecture validation code, but there is no harm in duplicating it.

Reply:
However, it may be useful there for other targets that implement PSA.

Reply:
For now I would keep as is.